File: js/src/days/day04.js
import { sum, reverseArray, skip } from "../lib/iterators.js"
import * as runner from "../lib/runner.js"
import { intersection } from '../lib/set.js'

console.log("Solving AoC 2023 day 04")

const readColumnedNumbers = (line, startIndex, untilChar, numberLength=2, dividerLength=1) => {
  const colWidth = numberLength + dividerLength
  const endIndex = untilChar ? line.indexOf(untilChar) - dividerLength : line.length
  const numbers = new Set()
  for (let i = startIndex; i < endIndex; i += colWidth) {
    const nextNumStr = line.substring(i, i + numberLength).trim()
    const nextNum = Number(nextNumStr)
    numbers.add(nextNum)
  }
  return numbers
}

const parseInput = (input) => {
  const lines = input.split('\n')
  const cards = lines.map((line) => {
    const winningNumbersStart = line.indexOf(":") + 2
    const winningNumbers = readColumnedNumbers(line, winningNumbersStart, "|")

    const myNumbersStart = line.indexOf("|") + 2
    const myNumbers = readColumnedNumbers(line, myNumbersStart)

    const matchedNumbers = intersection(winningNumbers, myNumbers)

    return {winningNumbers, myNumbers, matchedNumbers}
  })

  return cards
}

const getCardScore = ({matchedNumbers}) => {
  return matchedNumbers.size == 0 ? 0 : 2**(matchedNumbers.size - 1)
}

const part1 = (input) => {
  const cards = parseInput(input)
  const scores = cards.map(getCardScore)
  return sum(scores)
}

const part2 = (input) => {
  const cards = parseInput(input)
  const allCardsWon = [1]

  let curIndex = 1
  for (const card of skip(1, reverseArray(cards))) {
    let currentCardsWon = 1
    const numMatched = card.matchedNumbers.size
    for (let i = 1; i <= numMatched; i++) {
      currentCardsWon += allCardsWon[curIndex - i]
    }
    allCardsWon.push(currentCardsWon)
    curIndex++
  }

  return sum(allCardsWon)
}

await runner.testOutput('day04/example', '1', part1)
// await runner.printOutput('day04/test', part1)
// await runner.copyOutput('day04/test', part1)
// await runner.writeOutput('day04/test', '1', part1)
await runner.testOutput('day04/test', '1', part1)

await runner.testOutput('day04/example', '2', part2)
// await runner.printOutput('day04/test', part2)
// await runner.copyOutput('day04/test', part2)
// await runner.writeOutput('day04/test', '2', part2)
await runner.testOutput('day04/test', '2', part2)

File: cpp/src/days/day04.cpp
#import "day04.h"
#import "runner.h"
#import "string"
#include <cmath>
#include <numeric>
#include <set>
#include <sstream>
#include <vector>

namespace {

template <typename SetType> SetType intersectSets(const SetType &set1, const SetType &set2) {
    SetType result;
    std::set_intersection(set1.begin(), set1.end(), set2.begin(), set2.end(), std::inserter(result, result.begin()));
    return result;
}

struct Card {
    std::set<int> winningNumbers;
    std::set<int> myNumbers;
    std::set<int> matchedNumbers;
};

std::vector<Card> parseInput(const std::string &input) {
    std::istringstream iss(input);
    std::string line;
    std::vector<Card> cards;

    while (std::getline(iss, line)) {
        auto colonPos = line.find(':');
        auto pipePos = line.find('|');

        std::string winningNumbersStr = line.substr(colonPos + 2, pipePos - colonPos - 2);
        std::string myNumbersStr = line.substr(pipePos + 2);

        std::istringstream winNumStream(winningNumbersStr);
        std::istringstream myNumStream(myNumbersStr);

        std::set<int> winningNumbers;
        std::set<int> myNumbers;
        int num = 0;

        while (winNumStream >> num) {
            winningNumbers.insert(num);
        }

        while (myNumStream >> num) {
            myNumbers.insert(num);
        }

        std::set<int> matchedNumbers = intersectSets(winningNumbers, myNumbers);

        cards.emplace_back(Card{winningNumbers, myNumbers, matchedNumbers});
    }
    return cards;
}

int getCardScore(const std::set<int> &matchedNumbers) {
    return matchedNumbers.empty() ? 0 : static_cast<int>(std::pow(2, matchedNumbers.size() - 1));
}

std::string part1(const std::string &input) {
    auto cards = parseInput(input);
    int result = 0;
    for (const auto &card : cards) {
        result += getCardScore(card.matchedNumbers);
    }
    return std::to_string(result);
}

std::string part2(const std::string &input) {
    auto cards = parseInput(input);
    std::reverse(cards.begin(), cards.end()); // Reverse the vector to process in the correct order
    cards.erase(cards.begin());               // Drop the first element after reversing

    std::vector<int> allCardsWon;
    allCardsWon.push_back(1);

    int curIndex = 1;
    for (const auto &card : cards) {
        int currentCardsWon = 1;
        for (int i = 1; i <= card.matchedNumbers.size(); ++i) {
            currentCardsWon += allCardsWon[curIndex - i];
        }
        allCardsWon.push_back(currentCardsWon);
        curIndex++;
    }

    int result = std::accumulate(allCardsWon.begin(), allCardsWon.end(), 0);
    return std::to_string(result);
}

} // namespace

void day04() {
    testAocOutput("04", "example", "1", &part1);
    testAocOutput("04", "test", "1", &part1);
    testAocOutput("04", "example", "2", &part2);
    testAocOutput("04", "test", "2", &part2);
}
